Project backend plugin system

jdlesage opened the pull request:
## What changes are proposed in this pull request?

This pull request implements a plugin system to be able to propose new MLflow project backends.
It is a sort of RFC: there is doc modification, the new interface and the function to load a new backend (relying on python `entrypoint`)

It copies most of the logic of other MLflow plugins such as ArtifactStores, etc.

Databricks backend is the reference implemenation

## How is this patch tested?

1/ Install the mlflow-test-plugin project. It contains a "dummy" backend
2/ Go to examples/sklearn_elasticnet_wine and run `mlflow run --no-conda --backend dummy-backend .`
3/ It generates an entry in default experiment of MLflow. `'mlflow.project.backend` tag is  `dummy`

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

It introduces a plugin system to extend MLFlow Projects backends.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [X] Docs
- [ ] Tracking
- [X] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

jdlesage committed:
Add documentation

jdlesage committed:
Create plugin system

jdlesage committed:
Fix linting

jdlesage committed:
Remove databricks loader import

It will be for a next PR

jdlesage committed:
Add entrypoint group name

jdlesage committed:
Create the dummy plugin in example project

jdlesage committed:
Improve logging message

jdlesage commented:
Updated the "how is this patch tested?` with the dummy example

jdlesage committed:
remove extra space

jdlesage committed:
fix linter

jdlesage committed:
add databricks backend (reference implementation)

jdlesage committed:
add databricks files

jdlesage committed:
remove wrapping on 80 columns on databricks files

jdlesage committed:
Still remove some 80 columns wrapping

jdlesage committed:
remove last 80 columns wrap

jdlesage committed:
add backend to public symbols of mlflow.project

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,21 @@
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def run(self, active_run, uri, entry_point, work_dir, parameters,
```
Would you mind adding a docstring for this method describing the params & their types? 

Also, it'd be good to write out the possible options for this method signature & the pros/cons of the different options, e.g. another option could be to accept **kwargs here, but we might prefer that users just specify custom backend-specific arguments via `cluster_spec` (which BTW I think should be renamed to `backend_config`?)

It'd also be good to write out a high-level description of how this interface would work for the kubernetes & local backends, given that it excludes arguments like `storage_dir` that are passed to the local backend (maybe we just special-case such arguments & handle them before delegating to the plugin?

jdlesage replied:
done

jdlesage committed:
Document and clean AbstractBackend interface

Improve error message in loader

jdlesage committed:
Update the config.

Some parameters was used only by some backends. Therefore they are injected in the backend_config

jdlesage committed:
Factorize the method that generates env variables from run

jdlesage committed:
Add loader unit tests

jdlesage committed:
Fix call (parameters renaming)

jdlesage committed:
Break the plugin examples in part.

It fixes circular dependencies when entrypoints are loaded

jdlesage committed:
retrigger java tests

jdlesage committed:
Fix access to run_info in databrickcs backend

jdlesage committed:
If databricks validation failed then set status of run to FAILED

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,34 @@
+        :param uri: uri to the project (could be local or git uri).
+                    This is the parameter given to mlflow run command.
+        :param entry_point: name of the entry point to execute.
+        :param parameters: Dict of parameters to pass to the entry point
```
I wonder if we should keep `project_dir` (the local filesystem directory containing the fetched project) as a param here - it is common across backends & may not be obvious to plugin implementors otherwise that they don't actually need to fetch the project from `uri` in order to get access to its contents

jdlesage replied:
done

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,34 @@
+        """
+        Submit an entrypoint. It must returns a SubmittedRun object to track the execution
+
+        :param active_run: Current MLflow run object. Everything will registred in this run.
```
Could `active_run` be replaced with `run_id` containing just the ID of the MLflow run? And then to fetch the experiment ID, individual backends can call `run = mlflow.get_run(run_id); experiment_id = run.info.run_id`? Would be nice to keep the interface simpler for backend implementors (so that they don't necessarily need to learn the fields of `active_run`)

jdlesage replied:
done

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,34 @@
+        :param uri: uri to the project (could be local or git uri).
+                    This is the parameter given to mlflow run command.
+        :param entry_point: name of the entry point to execute.
+        :param parameters: Dict of parameters to pass to the entry point
```
```suggestion
        :param params: Dict of parameters to pass to the entry point
```

Just a small nit/suggestion on the argument name, for conciseness

jdlesage replied:
done

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,34 @@
+        Submit an entrypoint. It must returns a SubmittedRun object to track the execution
+
+        :param active_run: Current MLflow run object. Everything will registred in this run.
+        :param uri: uri to the project (could be local or git uri).
```
```suggestion
        :param project_uri: URI of the project (could be a local or git URI).
```

jdlesage replied:
done

smurching commented on mlflow/projects/backend/abstract_backend.py:
```diff
@@ -0,0 +1,34 @@
+        :param parameters: Dict of parameters to pass to the entry point
+        :param backend_config: Dict to pass parameters to the backend
+
+        :return: A SubmittedRun. This function is asynchronous, it should only trigger the run and
```
```suggestion
        :return: A :py:class:`mlflow.projects.SubmittedRun`. This function is expected to run the project asynchronously, i.e. it should trigger project execution and then immediately return a `SubmittedRun` to track execution status.
```

jdlesage committed:
Update mlflow/projects/backend/abstract_backend.py

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

jdlesage committed:
Fix interface.

Reintroduce project_dir

jdlesage committed:
Add validation backend config method

jdlesage committed:
Fetch and validation of project is performed by backend directly

Move to utils fetch and validate methods

jdlesage committed:
Start create fetch_and_validate method

jdlesage committed:
Merge remote-tracking branch 'mlflow/master' into project_backend_plugin

jdlesage committed:
Revert "Start create fetch_and_validate method"

This reverts commit 3002746faacce55d9885c29447786f4f73c6cc4e.

jdlesage committed:
Use id of the run that has been just created

jdlesage committed:
Fix & Move tests from test_projects to test_utils

jdlesage committed:
Move fixture tracking_uri_mock at project-level

jdlesage committed:
add missing import

jdlesage committed:
fix lint errors

jdlesage committed:
Remove warning unused argument

jdlesage committed:
typo

jdlesage committed:
remove import to mock tracking uri in tensorflow2 large test

smurching reviewed (commented):
Thanks for all the updates @jdlesage! One high-level comment, in order to modularize the changes & facilitate review, could we split this PR up into the following sub-PRs? I think they can be sequenced as follows:
   1. Adding the AbstractBackend interface & plugin infrastructure, refactoring the test plugin & updating plugin docs
  2. conftest.py changes (these are awesome btw! should save other developers a ton of time)
  3. Refactoring the Databricks backend to extend the AbstractBackend

I think 1 & 2 can actually be opened in parallel, with 3 to follow merging both of them (& after 3, we can merge an additional PR containing the refactoring you've already done of the k8s and local backends).

I'd expect 2 & 3 to be pretty easy to get merged, with 1 requiring the most (if any) iteration.

Thanks again for working on this! It'll be a great enhancement for MLflow & enable a ton of new use cases, and let me know if you'd like any help splitting up the PRs etc

aarondav commented:
@jdlesage Did you get a chance to look at @smurching's comments?

jdlesage commented:
Hi. I took a break last week... I didn't progress on this PR 

I will rebase the PR to the lastest version (to solve conflicts)  and take in account @smurching remarks asap.

jdlesage commented:
Hi @aarondav 
This PR has been split. The one that will be merged is https://github.com/mlflow/mlflow/pull/2566. It contains "only" the plugin part.

This PR is confusing, I close it.
